File: packages/sdk-node/src/snapshots.ts
import type { HttpClient } from "./http.js";
import type { SnapshotList, SnapshotRecord, Workspace } from "./types.js";

/**
 * Named snapshots — capture the current state of a workspace's output
 * directory and restore it into a new workspace later. Uses the engine's
 * freeze-before-copy path when a snapshot is taken mid-exec.
 *
 * ```ts
 * const snap = await aj.snapshots.create(ws.id, { name: "baseline" });
 * // …some risky exec…
 * const restored = await aj.snapshots.createWorkspaceFrom(snap.id);
 * // `restored.output_dir` now mirrors the baseline.
 * ```
 */
export class Snapshots {
  constructor(private readonly http: HttpClient) {}

  /**
   * Capture a snapshot of `workspace_id`'s output dir. If an exec is
   * currently running against the workspace, its cgroup is frozen for the
   * duration of the copy — callers see a consistent view.
   */
  async create(
    workspaceId: string,
    params: { name?: string } = {},
  ): Promise<SnapshotRecord> {
    const body: Record<string, unknown> = {};
    if (params.name !== undefined) body.name = params.name;
    return this.http.request<SnapshotRecord>({
      method: "POST",
      path: `/v1/workspaces/${encodeURIComponent(workspaceId)}/snapshot`,
      body,
    });
  }

  /** List snapshots; optionally filtered to a single workspace. */
  async list(params: {
    workspaceId?: string;
    limit?: number;
    offset?: number;
  } = {}): Promise<SnapshotList> {
    return this.http.request<SnapshotList>({
      method: "GET",
      path: "/v1/snapshots",
      query: {
        workspace_id: params.workspaceId,
        limit: params.limit,
        offset: params.offset,
      },
    });
  }

  /** Fetch a snapshot's metadata. */
  async get(id: string): Promise<SnapshotRecord> {
    return this.http.request<SnapshotRecord>({
      method: "GET",
      path: `/v1/snapshots/${encodeURIComponent(id)}`,
    });
  }

  /** Remove a snapshot + its on-disk dir. Idempotent. */
  async delete(id: string): Promise<void> {
    await this.http.request<void>({
      method: "DELETE",
      path: `/v1/snapshots/${encodeURIComponent(id)}`,
    });
  }

  /**
   * Rehydrate a snapshot into a brand-new workspace. The new workspace
   * inherits its parent's jail config (memory/network/etc) when the parent
   * is still around; otherwise sensible defaults apply.
   */
  async createWorkspaceFrom(
    snapshotId: string,
    params: { label?: string } = {},
  ): Promise<Workspace> {
    const body: Record<string, unknown> = { snapshot_id: snapshotId };
    if (params.label !== undefined) body.label = params.label;
    return this.http.request<Workspace>({
      method: "POST",
      path: "/v1/workspaces/from-snapshot",
      body,
    });
  }
}

File: packages/sdk-python/src/agentjail/snapshots.py
"""Snapshots — capture & restore workspace output dirs."""

from __future__ import annotations

from typing import Any

from ._http import HttpClient
from .types import SnapshotList, SnapshotManifest, SnapshotRecord, Workspace


class Snapshots:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def create(
        self,
        workspace_id: str,
        *,
        name: str | None = None,
    ) -> SnapshotRecord:
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        return self._http.request(
            "POST", f"/v1/workspaces/{workspace_id}/snapshot", json=body
        )

    def list(
        self,
        *,
        workspace_id: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
        q: str | None = None,
    ) -> SnapshotList:
        """List snapshots; optionally filtered to a workspace or by
        a substring search (``q``) matching ``id`` / ``name`` /
        ``workspace_id``.
        """
        return self._http.request(
            "GET",
            "/v1/snapshots",
            params={
                "workspace_id": workspace_id,
                "limit": limit,
                "offset": offset,
                "q": q,
            },
        )

    def get(self, snapshot_id: str) -> SnapshotRecord:
        return self._http.request("GET", f"/v1/snapshots/{snapshot_id}")

    def manifest(self, snapshot_id: str) -> SnapshotManifest:
        """List the files inside a pool-backed snapshot.

        Returns ``kind="incremental"`` with populated ``entries`` for
        snapshots captured into a content-addressed object pool;
        ``kind="classic"`` with empty ``entries`` for full-copy
        snapshots where the file list isn't persisted.
        """
        return self._http.request("GET", f"/v1/snapshots/{snapshot_id}/manifest")

    def delete(self, snapshot_id: str) -> None:
        self._http.request("DELETE", f"/v1/snapshots/{snapshot_id}")

    def create_workspace_from(
        self, snapshot_id: str, *, label: str | None = None
    ) -> Workspace:
        body: dict[str, Any] = {"snapshot_id": snapshot_id}
        if label is not None:
            body["label"] = label
        return self._http.request("POST", "/v1/workspaces/from-snapshot", json=body)
